EngineLayer/Calibration/CalibrationEngine.cs

```diff
@@ -1,332 +1,249 @@
-            var myXValueMatrix = new F64Matrix(myInputs.Count, myInputs.First().xValues.Length);
-            for (int i = 0; i < myInputs.Count; i++)
+        private double[] SmoothErrors(double[] relativeErrors, int numberOfScansUsedForSmoothingOnEachSide)
+        {
+            //impute missing values
+            //not all scans are guarenteed to contain data points. We can infer these data point with nearby points.
+            for (int index = 0; index < relativeErrors.Length; index++)
             {
-                for (int j = 0; j < myInputs.First().xValues.Length; j++)
+                if (relativeErrors[index] == 0) //if there were no points, then the value should be perfectly zero (the double default)
                 {
-                    myXValueMatrix[i, j] = myInputs[i].xValues[j];
+                    int startingBlankIndex = index;
+                    for (; index < relativeErrors.Length && relativeErrors[index] == 0; index++) { };
+                    double nextError = index == relativeErrors.Length ? relativeErrors[startingBlankIndex - 1] : relativeErrors[index]; //can't go all the way through without any data points, the original function checks for enough data points (where enough is more than zero)
+                    double previousError = startingBlankIndex > 0 ? relativeErrors[startingBlankIndex - 1] : nextError;
+                    int numberOfConsecutiveScansWithoutDataPoints = index - startingBlankIndex;
+                    for (int tempIndex = 0; tempIndex < numberOfConsecutiveScansWithoutDataPoints; tempIndex++)
+                    {
+                        relativeErrors[startingBlankIndex + tempIndex] = ((tempIndex + 1) * nextError + (numberOfConsecutiveScansWithoutDataPoints - tempIndex - 1) * previousError) / numberOfConsecutiveScansWithoutDataPoints;
+                    }
                 }
             }
 
-            var myYValues = myInputs.Select(p => p.yValue).ToArray();
-
-            // split data into training set and test set
-            var splitData = splitter.SplitSet(myXValueMatrix, myYValues);
-            var trainingSetX = splitData.TrainingSet.Observations;
-            var trainingSetY = splitData.TrainingSet.Targets;
-
-            // parameter ranges for the optimizer
-            var parameters = new ParameterBounds[]
+            //get correction factor for each scan, where half of the smooth comes from both directions
+            double[] smoothedErrors = new double[relativeErrors.Length];
+            int leftIndex = 0; //starting scan index used for numbers less than the current scan
+            int rightIndex = 1; //
+            double smoothedCorrectionFactor = 0;
+            //for scan #1 (index 0)
+            //no left scans, because we're at the beginning of the file. Just populate the first N number of scan errors
+            for (; rightIndex < numberOfScansUsedForSmoothingOnEachSide && rightIndex < relativeErrors.Length; rightIndex++)
```

Comment:
again here

Reply:
This loop is also needed, since we're getting the sum of all errors for the first "numberOfScansUsedForSmoothing". I'll put another commit with detailed comments. I incorrectly assumed it was an obvious implementation.

Reply:
A while loop would be cleaner here, too.